File: python/leetcode/my_pow.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        if n < 0:
            x, n = 1/x, -n
        if n == 0:
            return 1.0
        if x == 0:
            return 0
        r = 1.0
        while n:
            if n % 2:
                r *= x
            x *= x
            n //= 2
        return r

File: rust/leetcode/src/my_pow.rs
pub struct Solution;

impl Solution {
  pub fn my_pow(x: f64, n: i32) -> f64 {
    if n < 0 {
      return Solution::my_pow(1.0 / x, -n);
    }
    match x {
      0.0 => 0.0,
      1.0 => 1.0,
      _ => {
        let mut result = 1.0;
        let mut x = x;
        let mut n = n;
        while n != 0 {
          if n % 2 != 0 {
            result *= x;
          }
          x *= x;
          n >>= 1;
        }
        result
      }
    }
  }
}
